### PR TITLE
#162949070 Users should receive a JWT on successful Registration

### PR DESCRIPTION
**Pull request by waracci**
#### What does this PR do?

> Implement JWT upon user registration and add flake8 linting tool

#### Description of Task to be completed?

- Users should receive a `JSON Web Token` on successful Registration, the token is used to authenticate the user on protected routes.
- add flake8 linting tool to ensure all code submitted conforms to pep8 Style Guide for Python Code

#### How should this be manually tested?

1. Open terminal and clone the repo:

    `git clone https://github.com/andela/ah-robotics.git`

2. Change directory into the cloned repository.

    `cd ah-robotics`

3. Checkout to the Feature branch

    `git checkout ft-user-receive-jwt-162949070`

4. Create a virtual environment by running the following commands:

    `virtualenv -p python3 env`

    or

    `python3 -m virtualenv env`

5. Once the virtual environment has been setup, activate it and install the packages required by the project,
    these are contained in the file named `requirements.txt`

      `source env/bin/activate`

      `pip install -r requirements.txt`

### setting up the database

> Ensure PostgreSQL is installed and is running psql, follow these steps to create a new user and a database which will be used in the application.

  1. Start the postgres database server using the command:
  
      `pg_ctl -D /usr/local/var/postgres start`

  2. Run psql interactive terminal

      `psql postgres`

  3. Create a new user:

      `CREATE USER sample_username WITH PASSWORD 'sample_password';`

  4. Grant privileges to the user:

      `ALTER USER sample_username CREATEDB;`

  5. Create a database:

      `CREATE DATABASE sample_database_name WITH OWNER sample_username;`

  6. Exit psql interactive with the following command: `\q`

  7. Create a .env as guided by the .env_example and add the following configuration variables

      ```source-json
      source env/bin/activate
      export DB_NAME="database_name"
      export DB_USER="sample_username"
      export DB_PASS="sample_password"
      export DB_HOST="host_name_eg_localhost"
      export DB_PORT="port_number_eg_5432"
      export TIME_DELTA="30"
      ```

  7. Source .env to set the configuration variables:

      `source .env`
  8. Run migrations to create the tables in the database:

      ```source-json
      python manage.py makemigrations
      python manage.py migrate
      ```
---
 ### Run tests

   ```
python manage.py test

```
   

### Run The Server

Once everything has been setup, to run the application run the command:

   ```
python manage.py runserver

   ```

The application can be accessed in the url <http://localhost:8000/admin>

### Testing the feature

To test the feature, open [Postman](https://www.getpostman.com/) and input the following URL: <http://0.0.0.0:8000/api/users/>

To send a successful request, make sure to include a payload in the following format:
> NB: The payload should be placed in the `body` tab and ensure to select the `raw` option with the data type set to `JSON (application/json)`

       {
          "user" :{
           "username": "test_name",
           "email": "test_email@mail.com",
           "password": "password"
             }
        }

A successful response would be in the following format:

    {
        "user": {
        "email": "test_email@mail.com",
        "username": "test_name",
        "token": "b'eyJ0eXAiO9.eyJpZIiwidXNlcvlMJVUYYAbJSYOxW8F-bzhTBt2lk'"
        }
    }

---

#### What are the relevant pivotal tracker stories?

[#162949070](https://www.pivotaltracker.com/story/show/162949070)

#### Screenshots

> The correct payload format and expected output.

screenshot 2019-01-15 at 17 54 14

### CODE REVIEW

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @waracci @jomasim , 
Good work. It is running on my machine. However, please indicate in the Pull Request description that we should exit the psql terminal before we source our .env file.

Regards,
Paul Vitalis

**Comment by waracci** [comment]
> Hey @waracci @jomasim ,
> Good work. It is running on my machine. However, please indicate in the Pull Request description that we should exit the psql terminal before we source our .env file.
> 
> Regards,
> Paul Vitalis

well noted!

**Review by johnwayodi — APPROVED** [review]
Very good work. Works as expected.

**Comment by sharkdevs** [comment]
You have done a good work @waracci and @jomasim 
I only suggest that you add a `successful login` message as one of the fields in the output. It will help us give user feedback when consuming the API.

**Comment by Teatoller** [comment]
Dear @waracci 

I am able to follow the instructions 1-5 under "*How should this be manually tested?*". 
This bit is perfect and easy to follow.

I have also been able to follow the instructions for *setting up database* 1-5 successfully.
```
Database: authors
username: Teatoler
```
screenshot 2019-01-15 at 20 37 37

- Instruction 6 was not very clear but I want to think that I figured it out.
- The ```source .env``` was not able to start the activate the virtual environment for me when I ran it on the terminal.

screenshot 2019-01-15 at 20 23 48

Kindly check if something could be missing out or if a fix is needed to make it easier for an average user.

Regards
Steven Ennis

**Comment by jomasim** [comment]
> > > Dear @waracci
> > > I am able to follow the instructions 1-5 under "_How should this be manually tested?_".
> > > This bit is perfect and easy to follow.
> > > I have also been able to follow the instructions for _setting up database_ 1-5 successfully.
> > > ```
> > > Database: authors
> > > username: Teatoler
> > > ```
> > > screenshot 2019-01-15 at 20 37 37
> > > 
> > > * Instruction 6 was not very clear but I want to think that I figured it out.
> > > * The `source .env` was not able to start the activate the virtual environment for me when I ran it on the terminal.
> > > 
> > > screenshot 2019-01-15 at 20 23 48
> > > Kindly check if something could be missing out or if a fix is needed to make it easier for an average user.
> > > Regards
> > > Steven Ennis
> > 
> > 
> > @Teatoller On instruction 5 you were required to activate your virtualenv but it looks like you missed that as I can see on your terminal.  In order to activate your virtualenv run the following command :
> > ```
> > $ source env/bin/activate
> > ```
> > Then export all your variables by running
> > ```
> > $ source  .env
> > ```

**Comment by waracci** [comment]
> Dear @waracci
> 
> I am able to follow the instructions 1-5 under "_How should this be manually tested?_".
> This bit is perfect and easy to follow.
> 
> I have also been able to follow the instructions for _setting up database_ 1-5 successfully.
> 
> ```
> Database: authors
> username: teatoler
> ```
> screenshot 2019-01-15 at 20 37 37
> 
> * Instruction 6 was not very clear but I want to think that I figured it out.
> * The `source .env` was not able to start the activate the virtual environment for me when I ran it on the terminal.
> 
> screenshot 2019-01-15 at 20 23 48
> 
> Kindly check if something could be missing out or if a fix is needed to make it easier for an average user.
> 
> Regards
> Steven Ennis
Thanks Steve for your time to review and provide insightful feedback, we will take it into consideration.

**Comment by Collins33** [comment]
@waracci @jomasim I need more clarity on how to test the work you did. The instructions given are for just setting up the branch. More screenshots with the individual ways to actually test the feature for generating the jwt token. This is should be done after the server is started.

**Comment by waracci** [comment]
> @waracci @jomasim I need more clarity on how to test the work you did. The instructions given are for just setting up the branch. More screenshots with the individual ways to actually test the feature for generating the jwt token. This is should be done after the server is started.

Well noted, working on it.

**Comment by waracci** [comment]
> @waracci @jomasim I need more clarity on how to test the work you did. The instructions given are for just setting up the branch. More screenshots with the individual ways to actually test the feature for generating the jwt token. This is should be done after the server is started.

@Collins33, we have been able to make the requested changes. Kindly have a look.

**Comment by Paulvitalis200** [comment]
Hello @waracci @jomasim ,
Thank you for adding the postman requirement on the PR

Regards,
Paul Vitalis Otieno

**Comment by Teatoller** [comment]
Helle @waracci ,

Subsequent to my earlier feedback at the point ```7```  I suggest updating ```.env``` instructions by adding.
``` source env/bin/activate```
This way when the user runs the ```source .env``` on the terminal it will activate the virtual env.

screenshot 2019-01-16 at 13 04 11

The URL you have given http://localhost:8000/ for running the server gives the following image.
Consider suggesting for the user http://localhost:8000/admin/ just so that they don't land on a 404 page. See screenshot.

screenshot 2019-01-16 at 07 36 14

Lastly, I have been able to _successfully test in Postman_. **Good work!**.

screenshot 2019-01-16 at 13 48 20

**Comment by waracci** [comment]
> Hello @waracci ,
> 
> Subsequent to my earlier feedback at the point `7` I suggest updating `.env` instructions by adding.
> ` source env/bin/activate`
> This way when the user runs the `source .env` on the terminal it will activate the virtual env.
> 
> screenshot 2019-01-16 at 13 04 11
> 
> The url you have given http://localhost:8000/ for running the server gives the following image.
> Consider suggesting for the user http://localhost:8000/admin/ just so that the don't land on a 404 page. See screenshot.
> 
> screenshot 2019-01-16 at 07 36 14
> 
> Lastly, I have been able to _successfully test in Postman_. **Good work!**.
> 
> screenshot 2019-01-16 at 13 48 20

Thanks for the very elaborate review, your feedback is appreciated. 👍

**Comment by Collins33** [comment]
Awesome work on the implementation. Testing it on my local machine worked as expected. I, however, would like a small change, since export TIME_DELTA="30" is crucial to the working of the task, you could update the env. example and the readme file

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,67 @@
...
-# from django.conf import settings
-#
-# from rest_framework import authentication, exceptions
-#
-# from .models import User
+import jwt
+from django.conf import settings
+from rest_framework import exceptions
+from rest_framework.authentication import (
+    get_authorization_header, BaseAuthentication)
```
Comment (Collins33) [inline authors/apps/authentication/backends.py]: This file contains a lot of spaces. Remove them

**Review by Collins33 — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (waracci) [reply authors/apps/authentication/backends.py]: @Collins33  I got rid of the extra blank lines and run `pep8` checker on it, all checks are ok.

**Comment by sharkdevs** [comment]
It is good work especially on how much the tests are cover.

**Review by sharkdevs — COMMENTED** [review]
It is good work.

**Review by Collins33 — APPROVED** [review]
Nice work. The instructions are clear and it worked perfectly

**Pull request merged**